File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/MarkersController.java
package com.valentingrigorean.arcgis_maps_flutter.map;

import android.content.Context;

import com.esri.arcgisruntime.mapping.view.Graphic;
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay;
import com.valentingrigorean.arcgis_maps_flutter.Convert;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.flutter.plugin.common.MethodChannel;

public class MarkersController extends BaseSymbolController implements MapTouchGraphicDelegate {

    private final Map<String, MarkerController> markerIdToController = new HashMap<>();

    private final SymbolVisibilityFilterController symbolVisibilityFilterController;

    private final Context context;
    private final MethodChannel methodChannel;
    private final GraphicsOverlay graphicsOverlay;
    private final SelectionPropertiesHandler selectionPropertiesHandler;

    private MarkerController selectedMarker;


    public MarkersController(Context context, MethodChannel methodChannel, GraphicsOverlay graphicsOverlay, SelectionPropertiesHandler selectionPropertiesHandler, SymbolVisibilityFilterController symbolVisibilityFilterController) {
        this.context = context;
        this.methodChannel = methodChannel;
        this.graphicsOverlay = graphicsOverlay;
        this.selectionPropertiesHandler = selectionPropertiesHandler;
        this.symbolVisibilityFilterController = symbolVisibilityFilterController;
    }

    @Override
    public boolean canConsumeTaps() {
        for (final MarkerController controller : markerIdToController.values()) {
            if (controller.canConsumeTapEvents()) {
                return true;
            }
        }
        return false;
    }

    @Override
    public boolean didHandleGraphic(Graphic graphic) {
        final Object rawMarkerId = graphic.getAttributes().get("markerId");
        if (rawMarkerId == null) {
            return false;
        }
        final String markerId = (String) rawMarkerId;
        final MarkerController markerController = markerIdToController.get(markerId);
        if (markerController == null || !markerController.canConsumeTapEvents()) {
            return false;
        }
        if (selectedMarker != null)
            selectedMarker.setSelected(false);
        markerController.setSelected(true);
        selectedMarker = markerController;
        symbolVisibilityFilterController.invalidate(selectedMarker);
        methodChannel.invokeMethod("marker#onTap", Convert.markerIdToJson(markerId));
        return true;
    }

    public void addMarkers(List<Object> markersToAdd) {
        if (markersToAdd == null) {
            return;
        }
        executor.execute(() -> {
            for (Object marker : markersToAdd) {
                final Map<?, ?> data = (Map<?, ?>) marker;
                if (data == null) {
                    continue;
                }
                final String markerId = (String) data.get("markerId");
                final MarkerController markerController = new MarkerController(context, markerId, selectionPropertiesHandler);
                markerIdToController.put(markerId, markerController);
                Convert.interpretMarkerController(data, markerController, symbolVisibilityFilterController);

                addOrRemoveVisibilityFilter(symbolVisibilityFilterController, markerController, data);

                markerController.add(graphicsOverlay);
            }
        });
    }

    public void changeMarkers(List<Object> markersToChange) {
        if (markersToChange == null) {
            return;
        }
        executor.execute(() -> {
            for (Object marker : markersToChange) {
                final Map<?, ?> data = (Map<?, ?>) marker;
                if (data == null) {
                    continue;
                }
                final String markerId = (String) data.get("markerId");
                final MarkerController markerController = markerIdToController.get(markerId);
                if (markerController != null) {
                    Convert.interpretMarkerController(data, markerController, symbolVisibilityFilterController);
                    addOrRemoveVisibilityFilter(symbolVisibilityFilterController, markerController, data);
                }
            }
        });

    }

    public void removeMarkers(List<Object> markerIdsToRemove) {
        if (markerIdsToRemove == null) {
            return;
        }
        executor.execute(() -> {
            for (Object rawMarkerId : markerIdsToRemove) {
                if (rawMarkerId == null) {
                    continue;
                }
                final String markerId = (String) rawMarkerId;
                final MarkerController markerController = markerIdToController.remove(markerId);
                if (markerController != null) {
                    symbolVisibilityFilterController.removeGraphicsController(markerController);
                    markerController.remove(graphicsOverlay);
                }
            }
        });
    }

    public void clearSelectedMarker() {
        if (selectedMarker == null) {
            return;
        }
        selectedMarker.setSelected(false);
        symbolVisibilityFilterController.invalidate(selectedMarker);
        selectedMarker = null;
    }
}

File: android/src/main/java/com/valentingrigorean/arcgis_maps_flutter/map/MarkersController.kt
package com.valentingrigorean.arcgis_maps_flutter.map

import android.content.Context
import com.esri.arcgisruntime.mapping.view.Graphic
import com.esri.arcgisruntime.mapping.view.GraphicsOverlay
import com.valentingrigorean.arcgis_maps_flutter.Convert
import io.flutter.plugin.common.MethodChannel

class MarkersController(
    private val context: Context,
    private val methodChannel: MethodChannel,
    private val graphicsOverlay: GraphicsOverlay
) : BaseSymbolController(), MapTouchGraphicDelegate {
    private val markerIdToController: MutableMap<String?, MarkerController> = HashMap()
    private var selectedMarker: MarkerController? = null
    override fun canConsumeTaps(): Boolean {
        for (controller in markerIdToController.values) {
            if (controller.canConsumeTapEvents()) {
                return true
            }
        }
        return false
    }

    override fun didHandleGraphic(graphic: Graphic): Boolean {
        val rawMarkerId = graphic.attributes["markerId"] ?: return false
        val markerId = rawMarkerId as String
        val markerController = markerIdToController[markerId]
        if (markerController == null || !markerController.canConsumeTapEvents()) {
            return false
        }
        if (selectedMarker != null) selectedMarker!!.isSelected = false
        markerController.isSelected = true
        selectedMarker = markerController
        invalidateVisibilityFilterController(selectedMarker!!)
        methodChannel.invokeMethod("marker#onTap", Convert.Companion.markerIdToJson(markerId))
        return true
    }

    fun addMarkers(markersToAdd: List<Any?>?) {
        if (markersToAdd == null) {
            return
        }
        for (marker in markersToAdd) {
            val data = marker as Map<*, *>? ?: continue
            val markerId = data["markerId"] as String?
            val markerController = MarkerController(context, markerId)
            markerController.setSelectionPropertiesHandler(selectionPropertiesHandler)
            markerIdToController[markerId] = markerController
            Convert.Companion.interpretMarkerController(
                data,
                markerController,
                symbolVisibilityFilterController
            )
            markerController.add(graphicsOverlay)
        }
    }

    fun changeMarkers(markersToChange: List<Any?>?) {
        if (markersToChange == null) {
            return
        }
        for (marker in markersToChange) {
            val data = marker as Map<*, *>? ?: continue
            val markerId = data["markerId"] as String?
            val markerController = markerIdToController[markerId]
            if (markerController != null) {
                Convert.Companion.interpretMarkerController(
                    data,
                    markerController,
                    symbolVisibilityFilterController
                )
            }
        }
    }

    fun removeMarkers(markerIdsToRemove: List<Any?>?) {
        if (markerIdsToRemove == null) {
            return
        }
        for (rawMarkerId in markerIdsToRemove) {
            if (rawMarkerId == null) {
                continue
            }
            val markerId = rawMarkerId as String
            val markerController = markerIdToController.remove(markerId)
            if (markerController != null) {
                onSymbolRemoval(markerController)
                markerController.remove(graphicsOverlay)
            }
        }
    }

    fun clearSelectedMarker() {
        if (selectedMarker == null) {
            return
        }
        selectedMarker!!.isSelected = false
        invalidateVisibilityFilterController(selectedMarker!!)
        selectedMarker = null
    }
}